### ISSUE TITLE
Locking API wrappers are broken

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#21918]
In https://github.com/openssl/openssl/issues/20286#issuecomment-1701318469, @nicowilliams points out that the implementation of, at least, the pthread wrappers around pthreads are broken:
```
121 void ossl_crypto_mutex_lock(CRYPTO_MUTEX *mutex)
122 {
123     pthread_mutex_t *mutex_p;
124
125     mutex_p = (pthread_mutex_t *)mutex;
126     pthread_mutex_lock(mutex_p);
127 }
```

See that thread for the whole discussion, but here's a summary: if the only reason pthread can fail is because of openssl library misuse, then the library should assert. At least there should be a debug mode where that can happen. In general, `void` functions are a horrible idea because they swallow/hide the status of any internal failures. So even after the `ossl_assert` calls are added, the API must be changed to return a work/errors indicator.

**Comment by wtarreau** [comment]
Quite honestly, code that requires to check locks' return code has no chance to ever fly, because once you engage into this you generally have to imagine super complexe schemes to allow to roll back in case the impossible error happens and that often makes the whole code impossible. I'd rather suggest that these call an application-registered abort function if it ever happens (and it must not as you cannot recover from this if you have any lock held).

**Comment by nicowilliams** [comment]
As noted over on @20286, there are two cases where `pthread_mutex_lock()` can fail given how it is used in OpenSSL:

- `EDEADLK`, if there is a lock taking order bug in OpenSSL (perhaps there isn't any such _now_, but such a bug can always be introduced accidentally and not discovered until the code is in production);
- `EOWNERDEAD` -- I _think_ this could happen (maybe?) if the owning thread were cancelled, though I'm not sure.

**Comment by hlandau** [comment]
Copying over my comment since the discussion is moving here.

> > The only cases where acquiring a standard POSIX mutex can fail are cases which correspond to caller error which reflect a bug in our code in which an expected invariant is not being met.
> 
> There's two ways that POSIX mutexes can fail the way you're using them:
> 
>     1. you have a lock-taking order bug that leads to deadlocks that get detected
> 
>     2. a thread was canceled by the application or whatever (OpenSSL does not use `pthread_cleanup_{push,pop}()`
> 
> 
> Regarding (1), ignoring deadlock errors and continuing is arguably worse than having a deadlock to begin with.
Yes. We should assert.

> Regarding (2)... perhaps OpenSSL should use `pthread_cleanup_{push,pop}()`.

When has OpenSSL ever been safe for use with pthread_cancel? We cannot support
a use case where a thread executing OpenSSL library code might be terminated at
some arbitrary time while mutating internal OpenSSL data structures.

It also does not seem to me to be remotely worth augmenting the OpenSSL codebase
with these calls given the rather obscure and rare usage of `pthread_cancel`.
Of course, you're welcome to raise a PR if you think this is a valuable feature
for a use case you are interested in. But it seems like this is going to require
radical rearchitecture of basically all code in the OpenSSL codebase, whether or
not it uses locks.

>Can you explain which error codes from pthread_mutex_lock or pthread_mutex_unlock you are concerned about?
>
>EOWNERDEAD and EDEADLK.

EOWNERDEAD is not applicable because it applies only to robust mutexes used in
an interprocess fashion, neither of which are the case in OpenSSL.

EDEADLK is documented as follows:

       The pthread_mutex_lock() function may fail if:

       EDEADLK
              A deadlock condition was detected.

"may fail" — meaning this is UB and you have no guarantee about this error code
being returned, so it is not sound as a fault handling strategy. If it does
occur it indicates a bug and the only safe thing to do is to assert.

>What am I missing? I don't see how this is OK.

Yes. This is bad. We should assert here.

Regarding OpenSSL's excessive use of shared mutable state: wholly agree. This was an architectural mistake and we need to back out of it and find a way to do it that doesn't break API. So I'm basically in support of the architectural destination you're elaborating on here.

**Comment by nicowilliams** [comment]
It's true that OpenSSL can be declared not-safe-for-thread-cancellation, if it isn't already, and that, really, thread cancellation is fraught, and if you use it you get what you deserve.  Though apps that use thread cancellation may not be aware of what libraries are being used (again, see the good old name service switch w/o `nscd` care).  For example, a `libcurl`-using application that uses thread cancellation might be OK if `libcurl` is linked with one TLS library and not another.

So I'd be happy to leave `EOWNERDEAD` out of scope for now.

(BTW, even though the POSIX docs say that `EOWNERDEAD` is about robust shared locks, I have to believe that thread cancellation can also cause it to happen.  I've not checked various implementations yet, but it seems difficult to believe that thread cancellation would automatically drop held locks, and I can't find where that would be documented.)

**Comment by nicowilliams** [comment]
> EDEADLK is documented as follows:
> 
> ```
>    The pthread_mutex_lock() function may fail if:
> 
>    EDEADLK
>           A deadlock condition was detected.
> ```
> "may fail" — meaning this is UB and you have no guarantee about this error code
> being returned, so it is not sound as a fault handling strategy. If it does
> occur it indicates a bug and the only safe thing to do is to assert.

Eh?  No, that "may fail" does not mean that.  It means that `EDEADLK` is a possible error.

**Comment by hlandau** [comment]
I do not think we want to move to 'robust' locks &mdash; at least not without measuring the performance impact.

My point about `EDEADLK` is it doesn't guarantee it will do that &mdash; and above in the manpage it has a table which literally says relocking is "undefined behaviour". So you aren't reliably getting anything from it.

I will maintain that I think asserting is the only safe thing to do in the face of such an incident. It means we have a serious bug in our code and it is not at that point safe to assume anything.

**Comment by nicowilliams** [comment]
> I do not think we want to move to 'robust' locks — at least not without measuring the performance impact.
> 
> My point about `EDEADLK` is it doesn't guarantee it will do that — and above in the manpage it has a table which literally says relocking is "undefined behaviour". So you aren't reliably getting anything from it.

You are misreading the man page.  The table you're referring to is about what happens if you try to unlock a mutex you don't hold:

![image](https://github.com/openssl/openssl/assets/604851/df0d112f-d5bf-4d2b-9c18-62925156f4ef)
 
> I will maintain that I think asserting is the only safe thing to do in the face of such an incident. It means we have a serious bug in our code and it is not at that point safe to assume anything.

I can be convinced that this is true (though I'm not yet), but the code as it stands is definitely a bug.

**Comment by kroeckx** [comment]
Note that the windows equivalent of this function can't return an error

**Comment by nicowilliams** [comment]
> Note that the windows equivalent of this function can't return an error

I suppose one clever thing to do on `EDEADLK` is to just block forever, since that's what would have happened!  

```C
    for (;;) { struct timespec t = {1<<30, 0}; (void) nanosleep(&t, 0); }
```

I would consider that (along with maybe a warning message on `stderr`) an acceptable way to handle `EDEADLK`, and with that being the only relevant error, maybe that's good enough.

**Comment by paulidale** [comment]
Rather than a contrived "fix", we could change these calls to return an error code.  They are internal and are not subject to our API guarantee.

I'm also happy with the abort approach, since these are internal routines any misuse is a bug in our code.

**Comment by kroeckx** [comment]
So we have 2 sets of functions that wrap pthread locks:
- a public that uses rwlock when available, otherwise mutex
- a private doing mutex

Note that before commit e4d808652b0a1a19cfe615a6659e65ead0245108 the public mutex version was using NORMAL, now it's using DEFAULT, and so the relock cases also got changed to undefined behaviour on some platforms.

I think that we need to deprecate the public one.

It might also make sense to have 2 private ones, where 1 supports rwlocks, and 1 only does mutex. For some cases mutexes are better than rwlocks.

**Comment by hlandau** [comment]
> I think that we need to deprecate the public one.
> 
> It might also make sense to have 2 private ones, where 1 supports rwlocks, and 1 only does mutex. For some cases mutexes are better than rwlocks.
I agree with both of these points.

I previously discussed this at the Committers and OTC F2F regarding reformation to our locking APIs.

**Comment by nicowilliams** [comment]
> Err, "undefined behaviour" is clearly listed there under "default" for relocking also.

Sure, but you could use NORMAL non-robust locks.  Also, you can build an abstraction around POSIX mutexes that let you detect relock attempts.  (It's probably possible to build an abstraction around Windows mutexes that allow one to detect deadlocks too, though it might not perform well, and I've barely thought about it.)

> I don't view the error code approach as sane or safe. It is not too dissimilar from trying to mask SIGSEGV and make things "recoverable" in the event of a protection fault to make a program "resilient" to bugs. SIGSEGV should never happen but if it does, we must terminate the program as it is the only available safe course of action. See also libcs terminating on heap corruption etc. — it is the only safe recourse.

The default for `SIGSEGV` and `SIGBUS` is to kill the process, not to continue with UB.  If you like that then you should like `abort()`ing on lock error, and it sounds like you do:

> As I understand it, Illumos handles assertion failures in the kernel as panics, on the premise that if something has gone wrong with the kernel's invariants, the only safe thing to do is stop. This is based on Illumos having integrity as a primary value. By comparison Linux has its `OOPS` macro... which logs an assertion failure and then... continues running. Which is actually a pretty disturbing approach to quality and integrity of critical system software. I strongly support the Illumos approach here.

In this particular case there's also the option of blocking forever, which is what would happen on Windows.

**Comment by hlandau** [comment]
> The default for SIGSEGV and SIGBUS is to kill the process, not to continue with UB. If you like that then you should like abort()ing on lock error, and it sounds like you do:

Right, exactly. I support this.
